ENH: Do not wrap lines in logs (or provide an option to avoid wrapping)

In https://github.com/conda-forge/torchaudio-feedstock/pull/4, I see the following kind of lines
```
 │ │ │ =========================== short test summary info ============================
 │ │ │ FAILED test/torchaudio_unittest/compliance/kaldi/kaldi_compatibility_cpu_test.py::TestKaldiFloat32::test_fbank_00 - KeyError: 'foo\n\nTo execute this tes
 │ │ │ t, run the following from the base repo dir:\n     python test/run/etc/conda/test-files/torchaudio/1/test/torchaudio_unittest/compliance/kaldi/kaldi_comp
 │ │ │ atibility_cpu_test.py -k TestKaldiFloat32.test_fbank_00\n\nThis message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0'
 │ │ │ FAILED test/torchaudio_unittest/compliance/kaldi/kaldi_compatibility_cpu_test.py::TestKaldiFloat32::test_fbank_01 - KeyError: 'foo\n\nTo execute this tes
 │ │ │ t, run the following from the base repo dir:\n     python test/run/etc/conda/test-files/torchaudio/1/test/torchaudio_unittest/compliance/kaldi/kaldi_comp
 │ │ │ atibility_cpu_test.py -k TestKaldiFloat32.test_fbank_01\n\nThis message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0'
 │ │ │ FAILED [...]
```

And various other places where long lines get wrapped. This makes the logs very hard to read (compared to having all the test names nicely aligned, for example), so if not the default (my preference), then there should at least be an option to disable line wrapping.

Yeah I would argue though that on a terminal at home (that also does wrapping) that we have ugly empty lines on the left hand side. So it loooks like:

```
 │ │ │ FAILED test/torchaudio_unittest/compliance/kaldi/kaldi_compatibility_cpu_test.py::TestKaldiFloat32::test_fbank_00 - KeyError: 'foo\n\nTo execute this tes
t, run the following from the base repo dir:\n     python test/run/etc/conda/test-files/torchaudio/1/test/torchaudio_unittest/compliance/kaldi/kaldi_comp
atibility_cpu_test.py -k TestKaldiFloat32.test_fbank_00\n\nThis message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0'
 │ │ │ FAILED test/torchaudio_unittest/compliance/kaldi/kaldi_compatibility_cpu_test.py::TestKaldiFloat32::test_fbank_01 - KeyError: 'foo\n\nTo execute this tes
```

But I agree that it's annoying in CI so we can definitely automatically disable it there.

--- a/src/console_utils.rs
+++ b/src/console_utils.rs
@@ -297,16 +297,6 @@ impl Default for LoggingOutputHandler {
 }
 
 impl LoggingOutputHandler {
-    /// Create a new logging handler with the given multi-progress.
-    pub fn from_multi_progress(multi_progress: MultiProgress) -> LoggingOutputHandler {
-        Self {
-            wrap_lines: true,
-            state: Arc::new(Mutex::new(SharedState::default())),
-            progress_bars: multi_progress,
-            writer: io::stderr(),
-        }
-    }
-
     /// Return a string with the current indentation level (bars added to the
     /// front of the string).
     pub fn with_indent_levels(&self, template: &str) -> String {
@@ -318,6 +308,12 @@ impl LoggingOutputHandler {
     /// Return the multi-progress instance.
     pub fn multi_progress(&self) -> &MultiProgress {
         &self.progress_bars
+    }
+
+    /// Set the multi-progress instance.
+    pub fn with_multi_progress(mut self, multi_progress: MultiProgress) -> Self {
+        self.progress_bars = multi_progress;
+        self
     }
 
     /// Returns the style to use for a progressbar that is currently in

--- a/src/console_utils.rs
+++ b/src/console_utils.rs
@@ -1,7 +1,6 @@
 //! This module contains utilities for logging and progress bar handling.
 use std::{
     borrow::Cow,
-    collections::HashMap,
     future::Future,
     io,
     str::FromStr,
@@ -56,11 +55,17 @@ where
     }
 }
 
+#[derive(Debug)]
+struct SpanInfo {
+    id: Id,
+    start_time: Instant,
+    header: String,
+    header_printed: bool,
+}
+
 #[derive(Debug, Default)]
 struct SharedState {
-    indentation_level: usize,
-    timestamps: HashMap<Id, Instant>,
-    formatted_spans: HashMap<Id, String>,
+    span_stack: Vec<SpanInfo>,
     warnings: Vec<String>,
 }
 
@@ -155,104 +160,104 @@ where
     fn on_new_span(
         &self,
         attrs: &tracing_core::span::Attributes<'_>,
-        id: &tracing_core::span::Id,
+        id: &Id,
         ctx: Context<'_, S>,
     ) {
         let mut state = self.state.lock().unwrap();
-        state.timestamps.insert(id.clone(), Instant::now());
-        let span = ctx.span(id);
 
-        if let Some(span) = span {
+        if let Some(span) = ctx.span(id) {
             let mut s = Vec::new();
             let mut w = io::Cursor::new(&mut s);
             attrs.record(&mut CustomVisitor::new(&mut w));
             let s = String::from_utf8_lossy(w.get_ref());
 
-            if !s.is_empty() {
-                state
-                    .formatted_spans
-                    .insert(id.clone(), format!("{}{}", span.name(), s));
+            let name = if s.is_empty() {
+                span.name().to_string()
             } else {
-                state
-                    .formatted_spans
-                    .insert(id.clone(), span.name().to_string());
-            }
-        }
-    }
+                format!("{}{}", span.name(), s)
+            };
 
-    fn on_enter(&self, id: &Id, _ctx: Context<'_, S>) {
-        let mut state = self.state.lock().unwrap();
-        let ind = indent_levels(state.indentation_level);
-        if let Some(txt) = state.formatted_spans.get(id) {
-            eprintln!("{ind}\n{ind} {} {}", style("╭─").cyan(), txt);
-        }
+            let indent = indent_levels(state.span_stack.len());
+            let header = format!("{indent}\n{indent} {} {}", style("╭─").cyan(), name);
 
-        state.indentation_level += 1;
+            state.span_stack.push(SpanInfo {
+                id: id.clone(),
+                start_time: Instant::now(),
+                header,
+                header_printed: false,
+            });
+        }
     }
 
     fn on_exit(&self, id: &Id, _ctx: Context<'_, S>) {
         let mut state = self.state.lock().unwrap();
 
-        let prev_ind = indent_levels(state.indentation_level);
+        if let Some(pos) = state.span_stack.iter().position(|info| &info.id == id) {
+            let elapsed = state.span_stack[pos].start_time.elapsed();
+            let header_printed = state.span_stack[pos].header_printed;
+            state.span_stack.truncate(pos);
 
-        if state.indentation_level > 0 {
-            state.indentation_level -= 1;
+            if !header_printed {
+                return;
+            }
+
+            let indent = indent_levels(pos);
+            let indent_plus_one = indent_levels(pos + 1);
+
+            eprintln!(
+                "{indent_plus_one}\n{indent} {} (took {})",
+                style("╰───────────────────").cyan(),
+                HumanDuration(elapsed)
+            );
         }
-
-        let ind = indent_levels(state.indentation_level);
-
-        let elapsed_time = state
-            .timestamps
-            .remove(id)
-            .map(|t| t.elapsed())
-            .unwrap_or_default();
-
-        let human_duration = HumanDuration(elapsed_time);
-
-        eprintln!(
-            "{prev_ind}\n{ind} {} (took {})",
-            style("╰───────────────────").cyan(),
-            human_duration
-        );
     }
 
     fn on_event(&self, event: &Event<'_>, _ctx: Context<'_, S>) {
         let mut state = self.state.lock().unwrap();
-        let indent_str = indent_levels(state.indentation_level);
+        let indent = indent_levels(state.span_stack.len());
+
+        // Print pending headers
+        for span_info in &mut state.span_stack {
+            if !span_info.header_printed {
+                eprintln!("{}", span_info.header);
+                span_info.header_printed = true;
+            }
+        }
 
         let mut s = Vec::new();
         event.record(&mut CustomVisitor::new(&mut s));
-        let s = String::from_utf8_lossy(&s);
+        let message = String::from_utf8_lossy(&s);
 
-        let (prefix, prefix_len) =
-            if event.metadata().level() <= &tracing_core::metadata::Level::WARN {
-                state.warnings.push(s.to_string());
-                if event.metadata().level() == &tracing_core::metadata::Level::ERROR {
-                    (style("× error ").red().bold(), 7)
-                } else {
-                    (style("⚠ warning ").yellow().bold(), 9)
-                }
+        let (prefix, prefix_len) = if event.metadata().level() <= &Level::WARN {
+            state.warnings.push(message.to_string());
+            if event.metadata().level() == &Level::ERROR {
+                (style("× error ").red().bold(), 7)
             } else {
-                (style(""), 0)
-            };
-
-        let width: usize = terminal_size::terminal_size()
-            .map(|(w, _)| w.0)
-            .unwrap_or(160) as usize;
-
-        let max_width = width - (state.indentation_level * 2) - 1 - prefix_len;
+                (style("⚠ warning ").yellow().bold(), 9)
+            }
+        } else {
+            (style(""), 0)
+        };
 
         self.progress_bars.suspend(|| {
-            for line in s.lines() {
-                // split line into max_width chunks
-                if line.len() <= max_width {
-                    eprintln!("{} {}{}", indent_str, prefix, line);
-                } else {
-                    chunk_string_without_ansi(line, max_width)
-                        .iter()
-                        .for_each(|chunk| {
-                            eprintln!("{} {}{}", indent_str, prefix, chunk);
-                        });
+            if !self.wrap_lines {
+                for line in message.lines() {
+                    eprintln!("{} {}{}", indent, prefix, line);
+                }
+            } else {
+                let width = terminal_size::terminal_size()
+                    .map(|(w, _)| w.0)
+                    .unwrap_or(160) as usize;
+                let max_width = width - (state.span_stack.len() * 2) - 1 - prefix_len;
+
+                for line in message.lines() {
+                    if line.len() <= max_width {
+                        eprintln!("{} {}{}", indent, prefix, line);
+                    } else {
+                        for chunk in chunk_string_without_ansi(line, max_width) {
+                            eprintln!("{} {}{}", indent, prefix, chunk);
+                        }
+                    }
                 }
             }
         });
@@ -263,6 +268,7 @@ where
 #[derive(Debug)]
 pub struct LoggingOutputHandler {
     state: Arc<Mutex<SharedState>>,
+    wrap_lines: bool,
     progress_bars: MultiProgress,
     writer: io::Stderr,
 }
@@ -270,6 +276,7 @@ pub struct LoggingOutputHandler {
 impl Clone for LoggingOutputHandler {
     fn clone(&self) -> Self {
         Self {
+            wrap_lines: self.wrap_lines,
             state: self.state.clone(),
             progress_bars: self.progress_bars.clone(),
             writer: io::stderr(),
@@ -281,6 +288,7 @@ impl Default for LoggingOutputHandler {
     /// Creates a new output handler.
     fn default() -> Self {
         Self {
+            wrap_lines: true,
             state: Arc::new(Mutex::new(SharedState::default())),
             progress_bars: MultiProgress::new(),
             writer: io::stderr(),
@@ -292,6 +300,7 @@ impl LoggingOutputHandler {
     /// Create a new logging handler with the given multi-progress.
     pub fn from_multi_progress(multi_progress: MultiProgress) -> LoggingOutputHandler {
         Self {
+            wrap_lines: true,
             state: Arc::new(Mutex::new(SharedState::default())),
             progress_bars: multi_progress,
             writer: io::stderr(),
@@ -302,7 +311,7 @@ impl LoggingOutputHandler {
     /// front of the string).
     pub fn with_indent_levels(&self, template: &str) -> String {
         let state = self.state.lock().unwrap();
-        let indent_str = indent_levels(state.indentation_level);
+        let indent_str = indent_levels(state.span_stack.len());
         format!("{} {}", indent_str, template)
     }
 
@@ -482,6 +491,7 @@ impl<'a> MakeWriter<'a> for LoggingOutputHandler {
         self.clone()
     }
 }
+
 ///////////////////////
 // LOGGING CLI utils //
 ///////////////////////
@@ -560,11 +570,19 @@ pub fn init_logging(
     log_style: &LogStyle,
     verbosity: &Verbosity<InfoLevel>,
     color: &Color,
+    wrap_lines: Option<bool>,
     #[cfg(feature = "tui")] tui_log_sender: Option<
         tokio::sync::mpsc::UnboundedSender<crate::tui::event::Event>,
     >,
 ) -> Result<LoggingOutputHandler, ParseError> {
-    let log_handler = LoggingOutputHandler::default();
+    let mut log_handler = LoggingOutputHandler::default();
+
+    // Wrap lines by default, but disable it in CI
+    if let Some(wrap_lines) = wrap_lines {
+        log_handler.wrap_lines = wrap_lines;
+    } else if std::env::var("CI").is_ok() {
+        log_handler.wrap_lines = false;
+    }
 
     let use_colors = match color {
         Color::Always => Some(true),

--- a/src/main.rs
+++ b/src/main.rs
@@ -95,6 +95,7 @@ async fn main() -> miette::Result<()> {
                         &app.log_style,
                         &app.verbose,
                         &app.color,
+                        Some(true),
                         Some(tui.event_handler.sender.clone()),
                     )
                     .into_diagnostic()?;

--- a/src/main.rs
+++ b/src/main.rs
@@ -25,6 +25,7 @@ async fn main() -> miette::Result<()> {
                 &app.log_style,
                 &app.verbose,
                 &app.color,
+                app.wrap_log_lines,
                 #[cfg(feature = "tui")]
                 None,
             )

--- a/src/opt.rs
+++ b/src/opt.rs
@@ -129,6 +129,14 @@ pub struct App {
     )]
     pub log_style: LogStyle,
 
+    #[clap(
+        long,
+        env = "RATTLER_BUILD_WRAP_LOG_LINES",
+        default_value = "true",
+        global = true
+    )]
+    pub wrap_log_lines: Option<bool>,
+
     /// Enable or disable colored output from rattler-build.
     /// Also honors the `CLICOLOR` and `CLICOLOR_FORCE` environment variable.
     #[clap(

--- a/src/script.rs
+++ b/src/script.rs
@@ -374,16 +374,7 @@ impl Interpreter for CmdExeInterpreter {
 
         let build_script = format!(
             "{}\n{}",
-            CMDEXE_PREAMBLE
-                .replace("((script_path))", &build_env_path.to_string_lossy())
-                .replace(
-                    "((LIBRARY_INC))",
-                    &args.env_vars.get("LIBRARY_INC").unwrap_or(&"".to_string())
-                )
-                .replace(
-                    "((LIBRARY_LIB))",
-                    &args.env_vars.get("LIBRARY_LIB").unwrap_or(&"".to_string())
-                ),
+            CMDEXE_PREAMBLE.replace("((script_path))", &build_env_path.to_string_lossy()),
             args.script.script()
         );
         tokio::fs::write(

--- a/src/source/extract.rs
+++ b/src/source/extract.rs
@@ -214,8 +214,7 @@ mod test {
         let mut file = File::create(&file_path).unwrap();
         _ = file.write_all(HELLO_WORLD_ZIP_FILE);
 
-        let fancy_log = LoggingOutputHandler::default()
-            .with_multi_progress(multi_progress.clone());
+        let fancy_log = LoggingOutputHandler::default().with_multi_progress(multi_progress.clone());
 
         let res = extract_zip(file_path, tempdir.path(), &fancy_log);
         assert!(term.contents().trim().starts_with(

--- a/src/source/extract.rs
+++ b/src/source/extract.rs
@@ -214,7 +214,9 @@ mod test {
         let mut file = File::create(&file_path).unwrap();
         _ = file.write_all(HELLO_WORLD_ZIP_FILE);
 
-        let fancy_log = LoggingOutputHandler::from_multi_progress(multi_progress);
+        let fancy_log = LoggingOutputHandler::default()
+            .with_multi_progress(multi_progress.clone());
+
         let res = extract_zip(file_path, tempdir.path(), &fancy_log);
         assert!(term.contents().trim().starts_with(
             "Extracting zip       [00:00:00] [━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━]"

